### PR TITLE
[DFT] Remove comptime execution from DFT domain

### LINKED ISSUE #309: [DFT] Building multiple DFT backends at once causes linking conflicts
# Summary
When building multiple DFT backends at once, each backend defines a symbol for the many `oneapi::mkl::dft::compute_forward(...)` and `oneapi::mkl::dft::compute_backward(...)` compilation time dispatch overloads, so at linking time either one can be chosen.
To get around this I tried linking just one of the two backends, but then compilation failed because of an undefined reference to `oneapi::mkl::dft::detail::descriptor<...>::commit(oneapi::mkl::backend_selector<other_backend>)`. This is because both `ENABLE_MKLCPU_BACKEND`  and `ENABLE_MKLGPU_BACKEND` are defined in the preprocessor so both commits are defined here - 
https://github.com/oneapi-src/oneMKL/blob/develop/include/oneapi/mkl/dft/detail/descriptor_impl.hpp#L63 .
I tried getting around this by providing a definition of `commit` but made it a weak symbol so it can be overridden when the appropriate backend is linked. This led to problems when running the tests because https://github.com/oneapi-src/oneMKL/blob/develop/tests/unit_tests/main_test.cpp#L113 is still picking up both devices.

I believe other interfaces get around this by having a `backend_selector<backend::BACKEND>` parameter on all the functions, like here https://github.com/oneapi-src/oneMKL/blob/develop/include/oneapi/mkl/blas/detail/blas_ct_backends.hxx. I think the best course of action might be a spec change to include a `backend_selector` on `compute_forward` and `compute_backward`.

# Version
This on develop, I specifically tested githash 005c3c79a44974fc8c3adebfa71d94b0fe005f33

# Steps to reproduce
Build with both ENABLE_MKLCPU_BACKEND and ENABLE_MKLGPU_BACKEND set to 1.

# Observed behavior
When running `test_main_dft_ct` I get this output:
```
finlay@devpc-c310:~/perf/oneMKL/build$ ./bin/test_main_dft_ct 
Run this program with --terse_output to change the way it prints its output.
[==========] Running 488 tests from 3 test suites.
[----------] Global test environment set-up.
[----------] 240 tests from ComputeTestSuite/ComputeTests_in_place
[ RUN      ] ComputeTestSuite/ComputeTests_in_place.COMPLEX_SINGLE_in_place_buffer/sizes_8_batches_1_Intel_R__UHD_Graphics_750__0x4c8a_
Skipping test because: "oneMKL: DFT/compute_forward: function is not implemented Not implemented for MKLCPU"
...
```
# Expected behavior
The MKLGPU `compute_forward` should be called.

### PR DESCRIPTION
**Pull request by FMarno**
# Description

Due to the current design of the compile-time execution in the DFT domain, there can be redefinitions of the functions `compute_forward` and `compute_backward`, leading to unexpected behaviour when linking. This change removes the compile-time dispatch for `compute_forward` and `compute_backward` and also aligns the design more with the RNG domain. `commit` is not removed since it does not suffer from the same issues (due to it having a `backend_selector` argument).
The issue is discussed further in #309, and this patch is an alternative to #329.

Fixes #309

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log. 
- [x] Have you formatted the code using clang-format?

## Bug fixes

- [ ] Have you added relevant regression tests?
- [x] Have you included information on how to reproduce the issue (either in a
      GitHub issue or in this PR)?

### CODE REVIEW

**Comment by FMarno** [comment]
MKLGPU and MKLCPU run - 
[mklgpu_mklcpu_run.txt](https://github.com/oneapi-src/oneMKL/files/11733008/mklgpu_mklcpu_run.txt)

File: tests/unit_tests/include/test_helper.hpp [hunk 1]
```diff
@@ -172,6 +172,7 @@
                 TEST_RUN_NVIDIAGPU_CUBLAS_SELECT(q, func, __VA_ARGS__);    \
```
Comment (hjabird) [inline tests/unit_tests/include/test_helper.hpp]: Why are the changes in `test_helper.hpp` needed as part of this PR?

File: tests/unit_tests/dft/source/descriptor_tests.cpp [hunk 2]
```diff
@@ -431,8 +431,7 @@ inline void recommit_values(sycl::queue& sycl_queue) {
     std::vector<test_params> argument_groups{
         // not changeable
         // FORWARD_DOMAIN, PRECISION, DIMENSION, COMMIT_STATUS
-        { std::make_pair(config_param::NUMBER_OF_TRANSFORMS, std::int64_t{ 5 }),
```
Comment (hjabird) [inline tests/unit_tests/dft/source/descriptor_tests.cpp:434]: Why is this being deleted?

File: tests/unit_tests/dft/source/descriptor_tests.cpp [hunk 3]
```diff
@@ -538,8 +537,9 @@ inline void swap_out_dead_queue(sycl::queue& sycl_queue) {
     auto inout = sycl::malloc_device<forward_type>(default_1d_lengths + 2, sycl_queue);
     sycl_queue.wait();
 
-    auto transform_event = oneapi::mkl::dft::compute_forward<decltype(descriptor), forward_type>(
-        descriptor, inout, std::vector<sycl::event>{});
```
Comment (hjabird) [inline tests/unit_tests/dft/source/descriptor_tests.cpp]: Why did this need to change?

File: tests/unit_tests/dft/source/descriptor_tests.cpp (hunk 2 above)
Comment (FMarno) [reply tests/unit_tests/dft/source/descriptor_tests.cpp:434]: When number of transforms is > 1, mklcpu expects the FWD_DISTANCE and BWD_DISTANCE to be set appropriately, so this test was throwing an error saying "inconsistent configuration".

File: tests/unit_tests/include/test_helper.hpp (hunk 1 above)
Comment (FMarno) [reply tests/unit_tests/include/test_helper.hpp]: One of the changes wasn't helpful so I've removed it. The other changes removes a warning. Updated in https://github.com/oneapi-src/oneMKL/pull/331/commits/cccea27fa808f6d506ba3ab8b42072b587b7eb5d

File: tests/unit_tests/dft/source/descriptor_tests.cpp (hunk 3 above)
Comment (FMarno) [reply tests/unit_tests/dft/source/descriptor_tests.cpp]: reverted in https://github.com/oneapi-src/oneMKL/pull/331/commits/cccea27fa808f6d506ba3ab8b42072b587b7eb5d

File: src/dft/backends/mklcpu/CMakeLists.txt [hunk 4]
```diff
@@ -60,7 +59,7 @@ endif()
 set_target_properties(${LIB_OBJ} PROPERTIES
   POSITION_INDEPENDENT_CODE ON
 )
-target_link_libraries(${LIB_NAME} PUBLIC ${LIB_OBJ})
+target_link_libraries(${LIB_NAME} PUBLIC ${LIB_OBJ} onemkl_dft)
```
Comment (Rbiessy) [inline src/dft/backends/mklcpu/CMakeLists.txt:62]: From my current understanding of the CMake `onemkl_dft` should be linked to the tests and examples only to match with the other domains.

**Review by Rbiessy — COMMENTED** [review]
Thank you for this it looks good to me overall! Also could you clarify does this remove or affect any exported symbols? From what I can see these are only internal changes.

**Comment by FMarno** [comment]
> Thank you for this it looks good to me overall! Also could you clarify does this remove or affect any exported symbols? From what I can see these are only internal changes.

If I've got this all correct then the "function table" object in the dynamic library should be a lot smaller (no more compute_forward/backward), the `onemkl_dft_*****` libraries should no long have unique definitions of oneapi::mkl::dft::compute_forward/backward, but should export the ones from `onemkl_dft` (but I see you are suggesting removing that), and the `commit_impl` "interface" looks different but that shouldn't be exposed to the user.

File: src/dft/backends/mklcpu/CMakeLists.txt (hunk 4 above)
Comment (FMarno) [reply src/dft/backends/mklcpu/CMakeLists.txt:62]: Sorry, I'm having trouble seeing where this happens in the project. Saying that, I also don't see how the other domains are linking the runtime stuff at all.

Comment (FMarno) [reply src/dft/backends/mklcpu/CMakeLists.txt:62]: ~~I think that RNG might be doing their runtime dispatch in the headers~~ That's probably not completely true, but a lot of the work is happening in the headers.

Comment (Rbiessy) [reply src/dft/backends/mklcpu/CMakeLists.txt:62]: You're right, my bad. That looks good to me so let's keep it here!

**Review by Rbiessy — APPROVED** [review]

File: src/dft/backends/mklcpu/CMakeLists.txt (hunk 4 above)
Comment (Rbiessy) [reply src/dft/backends/mklcpu/CMakeLists.txt:62]: I think the `onemkl_${domain}` libraries may not be linked at all but opened in `dlopen` instead.

Comment (FMarno) [reply src/dft/backends/mklcpu/CMakeLists.txt:62]: I see! I think I know what I need to do but this is a little awkward with templates :D

**Comment by FMarno** [comment]
Closing in favour of #342

**Pull request closed without merging**